crates/bitwarden-crypto/src/cose.rs

```diff
@@ -15,22 +21,60 @@ use crate::{
 /// to be able to randomly generate nonces, and to not have to worry about key wearout. Since
 /// the draft was never published as an RFC, we use a private-use value for the algorithm.
 pub(crate) const XCHACHA20_POLY1305: i64 = -70000;
+const XCHACHA20_TEXT_PAD_BLOCK_SIZE: usize = 32;
+const CONTENT_TYPE_PADDED_UTF8: &str = "application/utf8-padded";
+
+/// The content format describes the format of the contained bytes. Message encryption always
+/// happens on the byte level, and this allows determining what format the contained data has. For
+/// instance, an `EncString` in most cases contains UTF-8 encoded text. In some cases it may contain
+/// a Pkcs8 private key, or a COSE key. Specifically, for COSE keys, this allows distinguishing
+/// between the old symmetric key format, represented as `ContentFormat::OctetStream`, and the new
+/// COSE key format, represented as `ContentFormat::CoseKey`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub enum ContentFormat {
+    Utf8,
+    Pkcs8,
+    CoseKey,
+    OctetStream,
+}
 
 /// Encrypts a plaintext message using XChaCha20Poly1305 and returns a COSE Encrypt0 message
 pub(crate) fn encrypt_xchacha20_poly1305(
     plaintext: &[u8],
     key: &crate::XChaCha20Poly1305Key,
+    content_format: ContentFormat,
 ) -> Result<Vec<u8>, CryptoError> {
-    let mut protected_header = coset::HeaderBuilder::new().build();
+    let protected_header = coset::HeaderBuilder::new();
+    let protected_header = match content_format {
+        // UTF-8 directly would leak the plaintext size. This is not acceptable for certain data
+        // (passwords).
+        ContentFormat::Utf8 => protected_header.content_type(CONTENT_TYPE_PADDED_UTF8.to_string()),
+        ContentFormat::Pkcs8 => protected_header.content_format(CoapContentFormat::Pkcs8),
+        ContentFormat::CoseKey => protected_header.content_format(CoapContentFormat::CoseKey),
+        ContentFormat::OctetStream => {
+            protected_header.content_format(CoapContentFormat::OctetStream)
+        }
+    };
+    let mut protected_header = protected_header.build();
     // This should be adjusted to use the builder pattern once implemented in coset.
     // The related coset upstream issue is:
     // https://github.com/google/coset/issues/105
     protected_header.alg = Some(coset::Algorithm::PrivateUse(XCHACHA20_POLY1305));
 
+    let encoded_plaintext = if content_format == ContentFormat::Utf8 {
```

Comment:
Question: Would it be better to move this up into the match on the content_format? This would protect us against accidentally only changing one in the future and having the content format mismatched from whether the contents are actually padded?

Reply:
Nice, changed to this split for the most part. I did not move the alg validation into the `TryFrom<&coset::Header>` implementation, because it does not really have much to do with the Content Format.